lib/active_model_serializers/adapter/json_api.rb

```diff
@@ -0,0 +1,186 @@
+      module ApiObjects
+        module JsonApi
+          ActiveModelSerializers.config.jsonapi_version = '1.0'
+          ActiveModelSerializers.config.jsonapi_toplevel_meta = {}
+          # Make JSON API top-level jsonapi member opt-in
+          # ref: http://jsonapi.org/format/#document-top-level
+          ActiveModelSerializers.config.jsonapi_include_toplevel_object = false
+
+          module_function
+
+          def add!(hash)
+            hash.merge!(object) if include_object?
+          end
+
+          def include_object?
+            ActiveModelSerializers.config.jsonapi_include_toplevel_object
+          end
+
+          # TODO: see if we can cache this
+          def object
+            object = {
+              jsonapi: {
+                version: ActiveModelSerializers.config.jsonapi_version,
+                meta: ActiveModelSerializers.config.jsonapi_toplevel_meta
+              }
+            }
+            object[:jsonapi].reject! { |_, v| v.blank? }
+
+            object
+          end
+        end
+      end
+
+      def initialize(serializer, options = {})
+        super
+        @include_tree = ActiveModel::Serializer::IncludeTree.from_include_args(options[:include])
+        @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options.delete(:fields))
+      end
+
+      def serializable_hash(options = nil)
+        options ||= {}
+
+        is_collection = serializer.respond_to?(:each)
+        serializers = is_collection ? serializer : [serializer]
+        primary_data, included = resource_objects_for(serializers)
+
+        hash = {}
+        hash[:data] = is_collection ? primary_data : primary_data[0]
+        hash[:included] = included if included.any?
+
+        ApiObjects::JsonApi.add!(hash)
+
+        if instance_options[:links]
+          hash[:links] ||= {}
+          hash[:links].update(instance_options[:links])
+        end
+
+        if is_collection && serializer.paginated?
+          hash[:links] ||= {}
+          hash[:links].update(pagination_links_for(serializer, options))
+        end
+
+        hash
+      end
+
+      def fragment_cache(cached_hash, non_cached_hash)
+        root = false if instance_options.include?(:include)
+        ActiveModelSerializers::Adapter::JsonApi::FragmentCache.new.fragment_cache(root, cached_hash, non_cached_hash)
+      end
+
+      protected
+
+      attr_reader :fieldset
+
+      private
+
+      def resource_objects_for(serializers)
+        @primary = []
+        @included = []
+        @resource_identifiers = Set.new
+        serializers.each { |serializer| process_resource(serializer, true) }
+        serializers.each { |serializer| process_relationships(serializer, @include_tree) }
+
+        [@primary, @included]
+      end
+
+      def process_resource(serializer, primary)
+        resource_identifier = ActiveModel::Serializer::Adapter::JsonApi::ApiObjects::ResourceIdentifier.new(serializer).as_json
```

Comment:
No, just is more scope for me :) ref https://github.com/rails-api/active_model_serializers/pull/1535#issuecomment-189119239 I'd rather put my energies elsewhere.  I was really just rebasing the original pr